src/Compression/CompressionCodecEncrypted.cpp

```diff
@@ -147,47 +143,120 @@ namespace DB
+        }
 
         assert(out_len == ciphertext.size() - 16);
     }
 
+    String unhexKey(const String & hex)
+    {
+        try
+        {
+            return boost::algorithm::unhex(hex);
+        }
+        catch (const std::exception &)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot read key_hex, check for valid characters [0-9a-fA-F] and length");
+        }
+    }
+
+    void CompressionCodecEncrypted::loadEncryptionKey(const Poco::Util::AbstractConfiguration & config, const String &config_prefix)
+    {
+        try
+        {
+            Strings config_keys;
+            config.keys(config_prefix, config_keys);
+            for (const std::string & config_key : config_keys)
+            {
+                String key;
+                UInt64 key_id;
+
+                if ((config_key == "key") || config_key.starts_with("key["))
+                {
+                    key = config.getString(config_prefix + "." + config_key, "");
+                    key_id = config.getUInt64(config_prefix + "." + config_key + "[@id]", 0);
+                }
+                else if ((config_key == "key_hex") || config_key.starts_with("key_hex["))
+                {
+                    key = unhexKey(config.getString(config_prefix + "." + config_key, ""));
+                    key_id = config.getUInt64(config_prefix + "." + config_key + "[@id]", 0);
+                }
+                else
+                    continue;
+
+                if (keys_storage.contains(key_id))
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Multiple keys have the same ID {}", key_id);
+                keys_storage[key_id] = key;
+            }
+
+            if (keys_storage.empty())
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "No keys, an encrypted disk needs keys to work");
+
+            current_key_id = config.getUInt64(config_prefix + ".current_key_id", 0);
+            if (!keys_storage.contains(current_key_id))
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Not found a key with the current ID {}", current_key_id);
+            if (keys_storage[current_key_id].size() != 16)
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS, "Got an encryption key with unexpected size {}, the size should be 16", keys_storage[current_key_id].size());
+            setMasterKey(keys_storage[current_key_id]);
+        }
+        catch (Exception & e)
+        {
+            e.addMessage("Exception in loadEncryptedKey.");
+            throw;
+        }
+    }
+
+    // Updates encryption keys from config.
+    void CompressionCodecEncrypted::updateEncryptionKeys(const Poco::Util::AbstractConfiguration & config, const String &config_prefix)
+    {
+        keys_storage.clear();
+        if (config.has("encryption_codecs.key_hex") || config.has("encryption_codecs.key"))
+            loadEncryptionKey(config, config_prefix);
```

Comment:
There can be other errors with new configuration, for example non-hex characters in `key_hex`. `try-catch` here would work better. And please rename the function to `bool tryLoad()`, that's how we normally call such things